proposals/20201026-creating-control-plane.md

```diff
@@ -0,0 +1,433 @@
+---
+title: Creating Control Plane
+authors:
+  - "@charleszheng44"
+reviewers:
+  - "@christopherhein"
+  - "@Fei-Guo"
+creation-date: 2020-10-26
+last-updated: 2020-11-09
+status: provisional
+---
+
+# Creating Control Plane
+
+## Table of Contents
+
+- [Title](#title)
+  - [Table of Contents](#table-of-contents)
+  - [Glossary](#glossary)
+  - [Summary](#summary)
+  - [Motivation](#motivation)
+    - [Goals](#goals)
+    - [Non-Goals/Future Work](#non-goalsfuture-work)
+  - [Proposal](#proposal)
+    - [Making NCPT more generic](#making-ncpt-more-generic)
+    - [Making NCP the rendezvous](#making-ncp-the-rendezvous)
+    - [Bootstrap](#bootstrap)
+    - [Creation](#creation)
+      - [Native way](#native-way)
+      - [Using the third-party provider](#using-the-third-party-provider)
+    - [Control Plane Custom Resources](#control-plane-custom-resources)
+      - [NestedEtcd CRD](#nestedetcd-crd)
+      - [NestedKAS CRD](#nestedkas-crd)
+      - [NestedKCM CRD](#nestedkcm-crd)
+    - [Security Model](#security-model)
+  - [Alternatives](#alternatives)
+  - [Implementation History](#implementation-history)
+
+## Glossary
+
+Refer to the [Cluster API Book Glossary](https://cluster-api.sigs.k8s.io/reference/glossary.html).
+
+The followings are terms that will be used in the proposal. Whether including them into the Glossary needs further discussion:
+
+* **Super Cluster** - The underlying cluster that manages the physical nodes, all pods created through the NestedControlPlanes will run on this cluster.
+* **NestedControlPlane(NCP)** - The control plane that are hosted on the super cluster. 
+* **NestedControlPlaneTemplate(NCPT)** - NCPT contains all information required to create the NCP.
+* **NestedEtcd(NEtcd)** - The etcd that belongs to the control plane of the nested cluster.
+* **NestedKAS(NKAS)** - The kube-apiserver which belongs to the control plane of the nested cluster.
```

Comment:
Yes, I agree. having `nested` sound verbose. Will rename the CRD

Reply:
Are we going to support non-K8s `APIServer` and `ControllerManager` as well?
And are we going to folk and customize these components?

If not, having a `kube` prefix looks verbose too and may indicate a different message to users/developers as well

Reply:
@brightzheng100 No, we only support the standard `apiserver` and `kcm`. Will change to `APIServer` and `ControllerManager`

Reply:
+1 to supporting only standard Kubernetes API Server and controller manager, that reduces the scope and keeps us in focus.

`Nested` prefix might be needed if we want to reuse the default API groups, e.g. `infrastructure.cluster.x-k8s.io` which RBAC is automatically setup for you.

Otherwise, if we want to use a different API group, and we drop the prefix, we need to setup RBAC in a way that CAPI manager can have permissions to act on these objects

Reply:
Thanks for that note @vincepri, that was my one concern. Nomenclature wise we can keep `Nested*` if that will cause issues.

Reply:
Agree @charleszheng44 @brightzheng100 `NestedAPIServer` `NestedControllerManager` and, `NestedEtcd`.